src/bin/nydus-image/merge.rs

```diff
@@ -0,0 +1,183 @@
+    fn from_meta(meta: &RafsSuperMeta) -> Self {
+        Self {
+            explicit_uidgid: meta.explicit_uidgid(),
+            has_xattr: meta.has_xattr(),
+            compressor: meta.get_compressor(),
+            digester: meta.get_digester(),
+        }
+    }
+
+    fn set_to_ctx(&self, ctx: &mut BuildContext) {
+        ctx.explicit_uidgid = self.explicit_uidgid;
+        ctx.has_xattr = self.has_xattr;
+        ctx.compressor = self.compressor;
+        ctx.digester = self.digester;
+    }
+}
+
+/// Merger merge multiple bootstraps (generally come from nydus tar blob of
+/// intermediate image layer) into one bootstrap (uses as final image layer).
+pub struct Merger {}
+
+impl Merger {
+    /// Merge assumes the bootstrap name as the hash of whole tar blob.
+    fn get_blob_hash(bootstrap_path: &Path) -> Result<String> {
+        let blob_hash = bootstrap_path
+            .file_name()
+            .ok_or_else(|| anyhow!("get file name"))?
+            .to_str()
+            .ok_or_else(|| anyhow!("convert to string"))?;
+        Ok(blob_hash.to_string())
+    }
+
+    pub fn merge(
+        ctx: &mut BuildContext,
+        sources: Vec<PathBuf>,
+        target: PathBuf,
+        chunk_dict: Option<PathBuf>,
+    ) -> Result<()> {
+        if sources.is_empty() {
+            bail!("please provide at least one source bootstrap path");
+        }
+
+        let mut tree: Option<Tree> = None;
+        let mut blob_mgr = BlobManager::new();
+        let mut flags: Option<Flags> = None;
+
+        for (layer_idx, bootstrap_path) in sources.iter().enumerate() {
+            // Get the blobs come from chunk dict bootstrap.
+            let mut chunk_dict_blobs = HashSet::new();
+            if let Some(chunk_dict_path) = &chunk_dict {
+                let rs = RafsSuper::load_from_metadata(&chunk_dict_path, RafsMode::Direct, true)
+                    .context(format!("load chunk dict bootstrap {:?}", bootstrap_path))?;
+                for blob in rs.superblock.get_blob_infos() {
+                    chunk_dict_blobs.insert(blob.blob_id().to_string());
+                }
+            }
+
+            let rs = RafsSuper::load_from_metadata(&bootstrap_path, RafsMode::Direct, true)
+                .context(format!("load bootstrap {:?}", bootstrap_path))?;
+
+            let current_flags = Flags::from_meta(&rs.meta);
+            if let Some(flags) = &flags {
+                if flags != &current_flags {
+                    bail!(
+                        "inconsistent flags between bootstraps, current bootstrap {:?}, flags {:?}",
+                        bootstrap_path,
+                        current_flags,
+                    );
+                }
+            } else {
+                // Keep final bootstrap following superblock flags of source bootstraps.
+                current_flags.set_to_ctx(ctx);
+                flags = Some(current_flags);
+            }
+
+            let parent_blobs = rs.superblock.get_blob_infos();
```

Comment:
Why not simply bail if `len(parent_blobs) > 1`? Then the `parent_blob_added` logic can be dropped.

Reply:
It is assumed that the `nydus-image create` at each layer and `nydus-image merge` commands use the same chunk dict bootstrap, so there is at most one blob from the parent bootstrap and all other blobs from the chunk dict bootstrap. Added a comment.

Reply:
Added more detailed comment.